02-KernelModules_CharacterDevice/driver.c

```diff
@@ -0,0 +1,223 @@
+{
+	printk("chrdev: size of buffer: %d\n", BUFFER_SIZE);
+	printk("chrdev: free: %d\n", (BUFFER_SIZE - data_size));
+	return 0;
+}
+
+static struct file_operations proc_file_fops = {
+	.owner = THIS_MODULE,
+	.read = proc_read
+};
+
+
+/*
+ * dev interface
+ */
+
+static int dev_open(struct inode *inodep, struct file *filep)
+{
+	if (1 == is_open) {
+		pr_err("chrdev: already open\n");
+		return -EBUSY;
+	}
+
+	is_open = 1;
+	pr_info("chrdev: device opened\n");
+	return 0;
+}
+
+static int dev_release(struct inode *inodep, struct file *filep)
+{
+	is_open = 0;
+	pr_info("chrdev: device closed\n");
+	return 0;
+}
+
+static ssize_t dev_read(struct file *filep, char *buffer, size_t len, loff_t *offset)
+{
+	int ret;
+
+	pr_info("chrdev: read from file %s\n", filep->f_path.dentry->d_iname);
+
+	if (len > data_size)
+		len = data_size;
+
+	ret = copy_to_user(buffer, data_buffer, len);
+	if (ret) {
+		pr_err("chrdev: copy_to_user failed: %d\n", ret);
+		return -EFAULT;
+	}
+	data_size = 0; /* eof for cat */
+
+	pr_info("chrdev: %ld bytes read\n", len);
+	return len;
+}
+
+static ssize_t dev_write(struct file *filep, const char *buffer, size_t len, loff_t *offset)
+{
+	int ret;
+
+	pr_info("chrdev: write to file %s\n", filep->f_path.dentry->d_iname);
+
+	data_size = len;
+	if (data_size > BUFFER_SIZE)
+		data_size = BUFFER_SIZE;
+
+	ret = copy_from_user(data_buffer, buffer, data_size);
+	if (ret) {
+		pr_err("chrdev: copy_from_user failed: %d\n", ret);
+		return -EFAULT;
+	}
+
+	pr_info("chrdev: %d bytes written\n", data_size);
+	return data_size;
+}
+
+struct file_operations fops = {
+	.owner = THIS_MODULE,
+	.open = dev_open,
+	.release = dev_release,
+	.read = dev_read,
+	.write = dev_write
+};
+
+
+static int __init device_init(void)
+{
+	is_open = 0;
+	data_size = 0;
+	error = 0;
+
+	if (BUFFER_SIZE < 1024)
+		BUFFER_SIZE = 1024;
+
+	data_buffer = kzalloc(BUFFER_SIZE * sizeof(*data_buffer), GFP_KERNEL);
+
+	major = register_chrdev(0, DEVICE_NAME, &fops);
+	if (major < 0) {
+		pr_err("register_chrdev failed: %d\n", major);
+		return major;
+	}
+	pr_info("chrdev: register_chrdev ok, major = %d\n", major);
+
+	pclass = class_create(THIS_MODULE, CLASS_NAME);
+	if (IS_ERR(pclass)) {
+		unregister_chrdev(major, DEVICE_NAME);
+		pr_err("chrdev: class_create failed\n");
+		return PTR_ERR(pclass);
+	}
+	pr_info("chrdev: device class created successfully\n");
+
+	pdev = device_create(pclass, NULL, MKDEV(major, 0), NULL, CLASS_NAME"0");
+	if (IS_ERR(pdev)) {
+		class_destroy(pclass);
+		unregister_chrdev(major, DEVICE_NAME);
+		pr_err("chrdev: device_create failed\n");
+		return PTR_ERR(pdev);
+	}
+
+	pr_info("chrdev: device node created successfully\n");
+
+	proc_file = proc_create(DEVICE_NAME, 0666, NULL, &proc_file_fops);
+
+	example_kobject = kobject_create_and_add(DEVICE_NAME, kernel_kobj);
+	if (!example_kobject)
+		return -ENOMEM;
+
+	error = sysfs_create_file(example_kobject, &sys_attribute.attr);
+
+    if (error)
```

Comment:
```suggestion
     if (error)
```

- WARNING: please, no spaces at the start of a line
Please fix code style
driver.c has style problems, please review.